### ISSUE TITLE
Define additional/custom hatches

### ISSUE THREAD

**Issue by mtb-za** [matplotlib/matplotlib#20690]
### Problem

The current set of hatching options within matplotlib is somewhat limited, being only the following, with the ability to make more dense ones.
![image](https://user-images.githubusercontent.com/1148889/126292084-467b893d-994d-41ae-bd56-9f065867f2f8.png)
In geoscience (and possibly other fields, but that is my background) there are a number of other hatching patterns that have long historical use and are still common, cf the USGS templates, available as a pdf [here](https://ngmdb.usgs.gov/fgdc_gds/geolsymstd/fgdc-geolsym-patternchart.pdf). Some of these are complex enough to be difficult to generate programmatically (eg 706: contorted schist), while others are much simpler (eg 328-K, which is a grid of offset `+`).

Is there a means to safely define hatches such the latter? A workaround, adding custom hatches to `_hatch_types` [can be done](https://github.com/agile-geoscience/striplog/blob/0c68f63d645c5bb7a5cc73b9bdaa197c4fb3cc33/striplog/hatches.py) but raises a deprecation warning (`MatplotlibDeprecationWarning: hatch must consist of a string of "*+-./OX\ox|" or None, but found the following invalid values "p". Passing invalid values is deprecated since 3.4 and will become an error two minor releases later.`) since PR https://github.com/matplotlib/matplotlib/pull/17926, and is therefore not a stable long-term solution.

### Proposed Solution

It would be good to have a means to define custom hatches in a safe way. In many cases I think it likely to be package maintainers adding the functionality (as `striplog` does) rather than users.

A less comprehensive solution that would still cover many common use-cases would be to allow for repetitive hatching of a single (user-chosen) symbol. (That is, being able to place a regular grid of `+` or `L` in the way that the current hatches API handles cases like `.` and `o`.)

**Comment by jklymak** [comment]
I don't think there would be a problem adding a public interface to add a patch pattern.  I think the largest issue would be documenting it. But if it was clear it was an advanced feature I don't imagine there would be a problem if you wanted to open a PR.

**Comment by mtb-za** [comment]
Thanks for the encouragement. I will have a think about how the API might look, and see if I can make something sensible for a PR.

**Comment by story645** [comment]
@mtb-za you can always start w/ a really rough PR and ask us to mark it as a draft if you wanna publicly hash it out.

**Comment by dopplershift** [comment]
You can also create a draft PR yourself by clicking the arrow to the right of the green "Create Pull Request" button (when submitting the PR) and selecting "Create draft pull request":
<img width="360" alt="image" src="https://user-images.githubusercontent.com/221526/126710764-008d477c-a9ad-45dd-8a87-47be6af3cede.png">

**Comment by tacaswell** [comment]
This is something we proposed in the most recent CZI proposal so it is reassuring to get further confirmation of user need :)

In the discussion around that @QuLogic said that any tiling path _should_ be able to be used as a hatch (including contorted schist).

**Comment by QuLogic** [comment]
Yes, theoretically it should be possible, but might need to do a bit of low-level surgery to expose it.

**Comment by MosGeo** [comment]
In addition to striplog, I believe projects like Cartopy will benefit greatly from this once geologic standards (e.g., https://github.com/cecgeology/USGS-FGDC) are implemented as matplotlib custom hatches.

**Comment by jorge243** [comment]
Hey! Any updates on this potentially very useful feature?

**Comment by timhoffm** [comment]
Obviously, no. Resources are limited and nobody has stepped up to take this on.

**Comment by jklymak** [comment]
I think this is on the roadmap for the coming year as part of the CZI grant, but not sure the status or who will do this work.  It is relatively major, though maybe just so far as it requires careful API decisions.

**Comment by dopplershift** [comment]
Here's a proposal: We don't perform the removal of the deprecation (#17926) until we have this API in place.

**Comment by dopplershift** [comment]
Actually, I'll go a step further and say don't do the removal until the release *after* the new API ships.

**Comment by timhoffm** [comment]
What's the advantage of continuing to accept invalid hatch patterns?

**Comment by anntzer** [comment]
It used to be possible to define custom hashes by fiddling with a private variable (_hatch_types), but that would not update validate_hatches to not warn on them (and because of the design of the hatch API (which is what it is...), it is not actually possible to infer what characters are valid, as that's hidden in the `HatchPatternSubclass.__init__` logic).  Private variable of course means unsupported, but the use case seems legitimate and we should not block it unless we have a replacement (or we deem the use case not legitimate).

**Comment by hjhilden** [comment]
Just pitching in to say that an update to the hatch API would benefit a lot in flexibility from including the possibility of direct control of orientation and contrast in addition to density: Textures or patterns are visually distinct if they differ sufficiently in their main three fundamental perceptual channel components (Ware 2013, _Information Visualization_  p. 202): orientation, scale and contrast.  Currently only scale (the density) can be changed easily. 

Not sure how this best might be achieved. Adobe Illustrator’s implementation of patterns uses clipping masks to deal with rotation.

**Comment by SBGit-2019** [comment]
I agree: Please keep the "unofficial" way working and maybe even remove the decrepation warning until there is an official API to support that feature.

**Comment by leejjoon** [comment]
These two packages use rectangular tiles containing a set of paths (clipped to a rectangle) to fill an area. Currently, this is implemented as a path effect, but it should be easy to adapt to the hatch API as far as it allows a callable object that takes the extent of the area to be filled as an argument. I think this is a simple but not very efficient solution as hatching is done by the artist not by the backend. On the other hand, if we would like to utilize hatching capability from the  backend side, things will get much more complicated.

For the former simpler approach, I can give it a try if there is an interest.

**Comment by timhoffm** [comment]
AFAIK we currently do hatching in the backend. I'm inclined to say that we should not add a different approach to the core library as that increases complexity. Is extended hatch configuration specialized enough so that using a third party library good enough? Can we do something to make hooking into the hatch easier for third parties without growing much complexity on our side?

**Comment by dopplershift** [comment]
Right now, while you can achieve the same effect as hatching using a tiling path effect, it's an entirely different parameter than passing in a value for the `hatch` parameter. IMO, until we have an API that allows users to extend the options for `hatch`, or we make `path_effects` the one well-documented way to accomplish hatching, this issue remains.

**Comment by timhoffm** [comment]
It seems all hatching is running through `Path.hatch()`.

https://github.com/matplotlib/matplotlib/blob/7ede0929dc39858a2801f8b6b2711db0d4fa5bfe/lib/matplotlib/path.py#L1029-L1037

which calls

https://github.com/matplotlib/matplotlib/blob/7ede0929dc39858a2801f8b6b2711db0d4fa5bfe/lib/matplotlib/hatch.py#L199-L204

So the backends effectively obtain a Path and individually decide how to render it.

If somebody is interested in looking into this: It should be possible to support other Paths. The main question is proably API and how we want to define the patterns. The string-based approach won't suffice here.

**Comment by leejjoon** [comment]
Here is a one approach.
This require very minimal change from the matplotlib side.

https://github.com/matplotlib/matplotlib/compare/main...leejjoon:matplotlib:custom-hatch#diff-9fa22d18bda986f54317196aa0414a09252e8421779c006ed9ac51bf229ed1e9

A user can simply set the hatch with a custom hatch instance. For example,

```python
from hatch_usgs import USGSPattern
patch.set_hatch(USGSPattern("605"))
```

![Image](https://github.com/user-attachments/assets/b23cc26a-cca9-4c4f-be6a-d6217623121d)

However, the current hatch implementation in Matplotlib is limited in the sense that the hatch pattern need to be designed in "a unit square".  It would be better to support pasterns of  different size, different width and height. The change for this need to be made every supported backends although I envision that the required change is not significant.

**Comment by Mat-Beut** [comment]
Since it hasn't been fetched yet to the main repo.
If we want to use it in our scripts, and store those scripts on github.
Do we have to clone the fork to import localy from here or can we download the USGpatterns and custom_hatch file, include it in our repo while crediting the main repo, fork repo and contributor that participated in the creation of those files ?

> Here is a one approach.
> This require very minimal change from the matplotlib side.
> 
> https://github.com/matplotlib/matplotlib/compare/main...leejjoon:matplotlib:custom-hatch#diff-9fa22d18bda986f54317196aa0414a09252e8421779c006ed9ac51bf229ed1e9
> 
> A user can simply set the hatch with a custom hatch instance. For example,
> 
> ```python
> from hatch_usgs import USGSPattern
> patch.set_hatch(USGSPattern("605"))
> ```
> 
> ![Image](https://github.com/user-attachments/assets/b23cc26a-cca9-4c4f-be6a-d6217623121d)
> 
> However, the current hatch implementation in Matplotlib is limited in the sense that the hatch pattern need to be designed in "a unit square".  It would be better to support pasterns of  different size, different width and height. The change for this need to be made every supported backends although I envision that the required change is not significant.
> 
> 
>

**Comment by story645** [comment]
@Mat-Beut I think you can just use the standalone library version if it fits your needs: https://mpl-pe-pattern-usgs.readthedocs.io/en/latest/
